admin/applications/__init__.py

```diff
@@ -60,14 +60,27 @@ def search():
             else:
                 data = application.application_data()
                 try:
-                    data = iter(data)
-                    if 'personal_details_data' in data:
-                        if 'first_name' in data.personal_details_data and search in data.personal_details_data.first_name.lower():
-                            applications.append(application)
-                        elif 'middle_names_or_empty_string' in data.personal_details_data and search in data.personal_details_data.middle_names_or_empty_string.lower():
-                            applications.append(application)
-                        elif 'last_name' in data.personal_details_data and search in data.personal_details_data.last_name.lower():
-                            applications.append(application)
+                    add_application = False
+
+                    if hasattr(data, 'personal_details_data'):
+                        if hasattr(data.personal_details_data, 'first_name') and search in data.personal_details_data.first_name.lower():
+                            add_application = True
+                        elif hasattr(data.personal_details_data, 'middle_names_or_empty_string') and search in data.personal_details_data.middle_names_or_empty_string.lower():
+                            add_application = True
+                        elif hasattr(data.personal_details_data, 'last_name') and search in data.personal_details_data.last_name.lower():
+                            add_application = True
+
+                    if hasattr(data, 'birth_registration_data'):
+                        if hasattr(data.birth_registration_data,'first_name' ) and search in data.birth_registration_data.first_name.lower():
+                            add_application = True
+                        elif hasattr(data.birth_registration_data, 'middle_names') and search in data.birth_registration_data.middle_names.lower():
+                            add_application = True
+                        elif hasattr(data.birth_registration_data, 'last_name') and search in data.birth_registration_data.last_name.lower():
+                            add_application = True
+
+                    if add_application:
+                        applications.append(application)
+
```

Comment:
With the new data structure, all sections are automatically created, so there's no need to use `hasattr(...)`.
It even seems happy if the string values are `None`.

So, this section can be simplified to be:
```python
data: ApplicationData = application.application_data()
try:
    add_application = (
        search in data.personal_details_data.first_name.lower() or
        search in data.personal_details_data.middle_names_or_empty_string.lower() or
        search in data.personal_details_data.last_name.lower() or
        search in data.birth_registration_data.first_name.lower() or
        search in data.birth_registration_data.middle_names.lower() or
        search in data.birth_registration_data.last_name.lower()
    )
    if add_application:
        applications.append(application)
except TypeError as e:
    pass
```

You'll need to add an import above to get the typing:
```python
from grc.business_logic.data_structures.application_data import ApplicationData
```

Reply:
Won't work, if the data is incomplete:
<img width="1753" alt="Screenshot 2023-02-08 at 11 40 10" src="https://user-images.githubusercontent.com/5941837/217520033-e2da631f-ce5a-4c38-a3ff-5bebf4b9cea0.png">

Reply:
Ooh that's interesting. It worked for me (but I was surprised that it did) - maybe I'm using a different version of python